### ISSUE TITLE
declarative pipeline: lock is engaged, even if stage's when condition does not apply

### ISSUE THREAD

**Issue by accumulator** [jenkinsci/lockable-resources-plugin#112]
I have defined a number of stages, contained in a parent stage, which also has a Lock defined in its `options` block.

In cases where the when condition does **not** apply, the lock is still engaged, causing the pipeline to wait for the lock to become available.

Showcase:

```
    stage('some stages') {
        when {
            expression {
                someCondition()
            }
        }
        stages {
            stage('stage1') {
            }
            stage('stage2') {
            }
        }
        options {
            lock(label: 'DEV', quantity: 1, variable: 'locked_env')
        }
    }
```

expected behaviour: The Lock is only requested if the stage's `when` condition evaluates to true

**Comment by abayer** [comment]
`options` are invoked before `when`. This is intentional.

**Comment by accumulator** [comment]
Ok. Is it then possible to pass the conditional to the lock option?

**Comment by dcendents** [comment]
That's why even though I'd like to switch to declarative pipelines I have to stick with scripted pipelines. I hope they do implement some better logic to wait for the lock before assigning expensive resources to the build (using an executor).

**Comment by accumulator** [comment]
I can understand why there would be technical reasons to invoke the `options` block before the `when`, for instance when the `when` condition needs some preparation. However it is indeed non-intuitive in this case.

Adding functionality to let the items in the option block declare themselves if they can be invoked ' early' or 'late' would be very welcome!

**Comment by abayer** [comment]
@accumulator Could you open a JIRA on issues.jenkins.io in the `pipeline-model-definition-plugin` component for adding a `beforeOptions` flag for `when`, like the existing `beforeAgent` flag? Thanks!

**Comment by famod** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-51865 contains a workaround (haven't tried it myself, yet).

@accumulator Did you create a ticket for `beforeOptions`?

**Comment by jimklimov** [comment]
Seems this would block my solution/workaround proposal for #134 :\

However, from the OP's example in that issue, I see that in at least scripted context the `lock` step can be used with a closure.

I did not try, but maybe that can be used in this case as well - have it used as a container of stages, parallels, steps etc. based on the needs and intention of a pipeline job developer? e.g.

````
    stage('some stages') {
        when {
            expression {
                someCondition()
            }
        }
        lock(label: 'DEV', quantity: 1, variable: 'locked_env') {
            stages {
               stage('stage1') {
                }
               stage('stage2') {
                }
            }
        }
    }
````

or

```
stage {
   steps {
      lock (...) {
         sh "make check"
      }
   }
}
````
and so on?

**Comment by TobiX** [comment]
I'm going to close this, since there is currently nothing I could do in the lockable-resources plugin which would change this behaviour. As @abayer said, this needs to be fixed in `pipeline-model-definition-plugin`. Feel free to reopen if the situation changes.

**Comment by famod** [comment]
FYI: Yesterday I created a PR against https://github.com/jenkinsci/pipeline-model-definition-plugin which introduces `beforeOptions` within `when`: https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/356

**Comment by famod** [comment]
FYI: Note: JENKINS-51865 is done, new `beforeOptions` in `when` is available in pipeline-model-definition 1.4.0.
